### ISSUE TITLE
Memory leak in PutMediaDecoder

### ISSUE THREAD

**Issue by bkempe** [aws/aws-sdk-java#1885]
We're seeing this memory leak warning using the `AmazonKinesisVideoPutMediaClient`:

```
LEAK: ByteBuf.release() was not called before it's garbage-collected. See http://netty.io/wiki/reference-counted-objects.html for more information.
Recent access records: 
Created at:
	io.netty.buffer.PooledByteBufAllocator.newDirectBuffer(PooledByteBufAllocator.java:331)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:185)
	io.netty.buffer.AbstractByteBufAllocator.directBuffer(AbstractByteBufAllocator.java:176)
	io.netty.buffer.AbstractByteBufAllocator.buffer(AbstractByteBufAllocator.java:113)
	io.netty.handler.codec.ByteToMessageDecoder.expandCumulation(ByteToMessageDecoder.java:529)
	io.netty.handler.codec.ByteToMessageDecoder$1.cumulate(ByteToMessageDecoder.java:89)
	com.amazonaws.services.kinesisvideo.internal.netty.handler.PutMediaDecoder.decode(PutMediaDecoder.java:116)
	com.amazonaws.services.kinesisvideo.internal.netty.handler.PutMediaDecoder.decode(PutMediaDecoder.java:43)
	io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:88)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	io.netty.channel.CombinedChannelDuplexHandler$DelegatingChannelHandlerContext.fireChannelRead(CombinedChannelDuplexHandler.java:438)
	io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
	io.netty.channel.CombinedChannelDuplexHandler.channelRead(CombinedChannelDuplexHandler.java:253)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	io.netty.handler.ssl.SslHandler.unwrap(SslHandler.java:1436)
	io.netty.handler.ssl.SslHandler.decodeJdkCompatible(SslHandler.java:1203)
	io.netty.handler.ssl.SslHandler.decode(SslHandler.java:1247)
	io.netty.handler.codec.ByteToMessageDecoder.decodeRemovalReentryProtection(ByteToMessageDecoder.java:502)
	io.netty.handler.codec.ByteToMessageDecoder.callDecode(ByteToMessageDecoder.java:441)
	io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:278)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
	io.netty.channel.DefaultChannelPipeline$HeadContext.channelRead(DefaultChannelPipeline.java:1434)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
	io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
	io.netty.channel.DefaultChannelPipeline.fireChannelRead(DefaultChannelPipeline.java:965)
	io.netty.channel.nio.AbstractNioByteChannel$NioByteUnsafe.read(AbstractNioByteChannel.java:163)
	io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:656)
	io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:591)
	io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:508)
	io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:470)
	io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:909)
	io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	java.lang.Thread.run(Thread.java:748)
```

FWIW, there appear to be some `try...finally...` missing around the usage of `ReferenceCountUtil.release` and the error handling code in https://github.com/aws/aws-sdk-java/blob/master/aws-java-sdk-kinesisvideo/src/main/java/com/amazonaws/services/kinesisvideo/internal/netty/handler/PutMediaDecoder.java, but we are not seeing an exception in this case.

**Comment by dagnir** [comment]
Hi @bkempe do you have a simple repro case we can use?

**Comment by dagnir** [comment]
Also please let us know what version of the SDK you're using.

**Comment by bkempe** [comment]
@dagnir The version is 1.11.415. This happens occasionally on a live system, but I've not been able to isolate it to a test case.

**Comment by debora-ito** [comment]
@bkempe Are you still experiencing the issue? 
Without a sample code it's really hard to reproduce it on our side.

**Comment by bkempe** [comment]
@debora-ito We are still seeing this. I will try to create a test to reproduce this.

**Comment by debora-ito** [comment]
Closing this due to inactivity, but please do reopen when you get a reproducible case.

**Comment by bkempe** [comment]
The issue was reproducible by running https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-java/blob/master/src/main/demo/com/amazonaws/kinesisvideo/demoapp/PutMediaDemo.java for 1 minute using SDK 1.11.415, but seems to be fixed when using 1.11.487 or greater. Thanks!

**Comment by bkempe** [comment]
I still have seen the problem with the latest SDK 1.11.519, but so far haven't been able to reproduce it with PutMediaDemo.
